Update container orchestration for development environment...

ivan-junior-rh opened the pull request:
Update container orchestration for development environment for backend execution isolated from frontend.

# Features
- [X] Allow to run the backend isolated from frontend;
- [X] Remove `dockerize` dependency;
- [X] Add `config.py` file for `gunicorn` customization;
- [X] Optionally, avoid `python`/`pip` dependency locally;
- [X] Optionally, avoid `packages` installation locally;
- [X] Allow to build the images in `docker way` (without using `make`);
- [X] Additional customization for bind address and port via env variables.

# Remarks:
- The `RHUB_RETURN_INITIAL_FLASK_APP` flag was add to allow `flask routes` in `docker build`, without dealing with the dependency injection (we can create another file to start a simple flask app for this, but I don't think the work would worth it).
- The boolean values were changed to `True`/`False`, so we can casting them easily in python code;

ivan-junior-rh commented:
@daniellutz suggested to use `postgres` for `keycloak` data and I think that is a good idea. Moving from `h2` to `postgresql` requires some additional data for database connection.

I just pushed an additional commit, adding a `RHUB` prefix for database connection vars, since `keycloak` uses `DB_` prefix already, and I didn't want to put everything into the same database to avoid collections mess. Now, we can have both `rhub` and `keycloak` databases in postgres.

I would like to know your thoughts about this approach...

brhatiga commented:
Using a proper DB in production was always the plan so doing so in our dev env makes perfect sense to me since it will more closely mimic production.  +1

fpob commented on Dockerfile:
```diff
@@ -1,18 +1,28 @@
 
+WORKDIR /opt/app-root/src/
+
+RUN set -a; source ./.env.defaults; set +a; \
```
This was just a simple sanity check in makefile to make sure the container was built ok. I don't think it should be there...

fpob replied:
And FYI, `.env` may look like it is bash and it is safe to source it, but it's not. In `.env` files are no special characters, and quotes are not supported, they are part of value. `FOO=;:(){ :|:& };:` is ok in `.env`, but you don't want to source it into bash, I guess :smile:

ivan-junior-rh replied:
I was sourcing the `.defaults` file, with only keys and no values, so our bash is still working :)

The idea was to avoid exception raise in `_config.py` file.

However, the right way (imho) of doing that is to change in `_config.py` the `os.environ` to `os.getenv` with a default value (`''`), so we would not get an exception when building the image (OTOH we'll not raise this exception when running the application, but I don't think this is a problem).

Do you think this is okay? (I already force-pushed the changes in `_config.py`).

fpob replied:
`getenv` is already used for all other env variables anyway, so I think it's ok.

fpob commented on Makefile:
```diff
@@ -5,23 +5,26 @@ init:
+	docker-compose down
+
+test:
+	pytest -s tests/
```
`tox`

ivan-junior-rh replied:
Done, thanks!

fpob commented on docker-compose.yml:
```diff
@@ -1,24 +1,57 @@
+
+volumes:
+  postgres_data:
+  rhub-api:
```
`rhub-api` and `rhub-app` volumes are not used anywhere...

ivan-junior-rh replied:
True. Removed.

fpob commented on docker-compose.yml:
```diff
@@ -1,24 +1,57 @@
+      - postgres_data:/var/lib/postgresql/data
+      - ./bin/setup_postgres.sh:/docker-entrypoint-initdb.d/setup_postgres.sh:delegated,z
+
+networks:
```
Duplicate, `networks` is already at the beginning of the file. This silently overrides the original correct value, and containers are not connected to `rhub-development` network as they should.

ivan-junior-rh replied:
Fixed.

fpob commented on docker-compose.yml:
```diff
@@ -1,24 +1,52 @@
+    env_file:
+      - .env
+    ports:
+      - 5000:8081
```
wrong port, 8081 is default, not 5000

```suggestion
      - 8081:8081
```

ivan-junior-rh replied:
Hmmm, I think we were using different ports in the host because `8080` from frontend was conflicting with `keycloak` port. Anyway, since we don't have anything using `8081` in the host, I made the change...

ivan-junior-rh commented:
In the latest `push` I rebased from `master` and added `RHUB_UBI_PYTHON_IMAGE` to `Dockerfile`, so we can define a specific UBI python version via `--build-arg`...

fpob commented:
I don't think `RHUB_UBI_PYTHON_IMAGE` will be useful for something other than breaking the container... python36 won't work, python37 won't work too, ...

ivan-junior-rh commented on README.md:
````diff
@@ -4,18 +4,33 @@ Resource Hub API/backend service
+docker-compose up  # starts the api
+```
+
+Additionaly, for conveniency, there is a `Makefile` with some useful commands:
````
Done. Thanks!

cbouchar commented on .env.defaults:
```diff
@@ -1,39 +1,66 @@
+################################################################################
+FLASK_APP=
+RHUB_SKIP_INIT=
+#RHUB_RETURN_INITIAL_FLASK_APP=
```
Did you want this commented out?

ivan-junior-rh replied:
Yes :)
I removed it, since it's not necessary anymore at this point...

ivan-junior-rh commented:
> I don't think `RHUB_UBI_PYTHON_IMAGE` will be useful for something other than breaking the container... python36 won't work, python37 won't work too, ...

OK, removed it. If we need it in the future, we can discuss about it.
